gcore/gdalrasterband.cpp

```diff
@@ -1929,6 +1929,128 @@ uint64_t GDALRasterBand::GetNoDataValueAsUInt64(int *pbSuccess)
+        else
+            dfReplacementVal = dfNoDataValue + 1;
+    }
+    else if (dt == GDT_Int8)
+    {
+        dfNoDataValue = GDALClampDoubleValue(
+            dfNoDataValue, std::numeric_limits<int8_t>::lowest(),
+            std::numeric_limits<int8_t>::max());
+        if (dfNoDataValue == std::numeric_limits<GInt8>::max())
+            dfReplacementVal = std::numeric_limits<GInt8>::max() - 1;
+        else
+            dfReplacementVal = dfNoDataValue + 1;
+    }
+    else if (dt == GDT_UInt16)
+    {
+        dfNoDataValue = GDALClampDoubleValue(
+            dfNoDataValue, std::numeric_limits<uint16_t>::lowest(),
+            std::numeric_limits<uint16_t>::max());
+        if (dfNoDataValue == std::numeric_limits<GUInt16>::max())
+            dfReplacementVal = std::numeric_limits<GUInt16>::max() - 1;
+        else
+            dfReplacementVal = dfNoDataValue + 1;
+    }
+    else if (dt == GDT_Int16)
+    {
+        dfNoDataValue = GDALClampDoubleValue(
+            dfNoDataValue, std::numeric_limits<int16_t>::lowest(),
+            std::numeric_limits<int16_t>::max());
+        if (dfNoDataValue == std::numeric_limits<GInt16>::max())
+            dfReplacementVal = std::numeric_limits<GInt16>::max() - 1;
+        else
+            dfReplacementVal = dfNoDataValue + 1;
+    }
+    else if (dt == GDT_UInt32)
+    {
+        dfNoDataValue = GDALClampDoubleValue(
+            dfNoDataValue, std::numeric_limits<uint32_t>::lowest(),
+            std::numeric_limits<uint32_t>::max());
+        if (dfNoDataValue == std::numeric_limits<GUInt32>::max())
+            dfReplacementVal = std::numeric_limits<GUInt32>::max() - 1;
+        else
+            dfReplacementVal = dfNoDataValue + 1;
+    }
+    else if (dt == GDT_Int32)
+    {
+        dfNoDataValue = GDALClampDoubleValue(
+            dfNoDataValue, std::numeric_limits<int32_t>::lowest(),
+            std::numeric_limits<int32_t>::max());
+        if (dfNoDataValue == std::numeric_limits<GInt32>::max())
+            dfReplacementVal = std::numeric_limits<GInt32>::max() - 1;
+        else
+            dfReplacementVal = dfNoDataValue + 1;
+    }
+    else if (dt == GDT_Float32)
+    {
+
+        dfNoDataValue = GDALClampDoubleValue(
+            dfNoDataValue, std::numeric_limits<float>::lowest(),
+            std::numeric_limits<float>::max());
+
+        const float fNoDataValue{static_cast<float>(dfNoDataValue)};
+
+        if (fNoDataValue == std::numeric_limits<float>::infinity())
+        {
+            dfNoDataValue = std::numeric_limits<float>::max();
+        }
+        else if (fNoDataValue == -std::numeric_limits<float>::infinity())
+        {
+            dfNoDataValue = std::numeric_limits<float>::lowest();
+        }
+
+        if (dfNoDataValue == std::numeric_limits<float>::max())
+        {
+            dfReplacementVal = std::nextafter(
+                dfNoDataValue, std::numeric_limits<float>::lowest());
```

Comment:
I woud rather take the nextbefore() of max(), as we want to minimize the gap between dfNoDataValue and dfReplacementVal

Reply:
> I woud rather take the nextbefore() of max(), as we want to minimize the gap between dfNoDataValue and dfReplacementVal

To my knowledge nextbefore does not exist, nextafter does the job: https://cpp.sh/?source=%2F%2F+Example+program%0A%23include+%3Ciostream%3E%0A%23include+%3Cstring%3E%0A%23include+%3Cmemory%3E%0A%0Astruct+A%0A%7B%0A++++int+foo+%3D+1%3B%0A%7D%3B%0A%0Astruct+B%0A%7B%0A++++B(std%3A%3Aunique_ptr%3CA%3E+%26%26_a)%3A%0A+++++a(std%3A%3Amove(_a))+%7B%0A+++++++++%0A++++std%3A%3Acout+%3C%3C+%22a%3A+%22+%3C%3C+a+%3C%3C+std%3A%3Aendl%3B%0A++++std%3A%3Acout+%3C%3C+%22_a%3A+%22+%3C%3C+_a+%3C%3C+std%3A%3Aendl%3B%0A%7D%3B%0A+++++%0A++++std%3A%3Aunique_ptr%3CA%3E+a%3B%0A%7D%3B%0A%0Aint+main()%0A%7B%0A%0A++++std%3A%3Aunique_ptr%3CA%3E+a+%7B+new+A(+)+%7D%3B%0A++++a-%3Efoo+%3D+2%3B%0A++++%0A++++std%3A%3Acout+%3C%3C+%22a%3A+%22+%3C%3C+a+%3C%3C+std%3A%3Aendl%3B%0A++++%0A++++B+b%7Bstd%3A%3Amove(a)%7D%3B%0A++++%0A++++std%3A%3Acout+%3C%3C+%22a%3A+%22+%3C%3C+a+%3C%3C+std%3A%3Aendl%3B%0A%7D

What I will do is replace max/lowest as the second argument with 0 when appropriate to make it more readable.